File: fixtures/scale/dart/large/lib/ciphers/xtea/xtea_cipher.dart
import '../../core/cipher.dart';
import '../../util/block_codec.dart';
import '../../util/bytes.dart';
import '../../util/hex.dart';
import 'xtea_key.dart';
import 'xtea_rounds.dart';

class XteaCipher implements Cipher {
  final XteaKey key;

  XteaCipher(this.key);

  @override
  String name() => 'xtea';

  @override
  String encrypt(String plaintext) {
    final data = padBytes(bytesOf(plaintext), 8);
    for (var off = 0; off < data.length; off += 8) {
      writeBlock(data, off, encryptBlock(readBlock(data, off), key));
    }
    return hexEncode(data);
  }

  @override
  String decrypt(String ciphertext) {
    final data = hexDecode(ciphertext);
    for (var off = 0; off < data.length; off += 8) {
      writeBlock(data, off, decryptBlock(readBlock(data, off), key));
    }
    return bytesToText(data).trim();
  }
}

File: fixtures/scale/scala/large/src/cryptobench/ciphers/xtea/XteaCipher.scala
package cryptobench.ciphers.xtea

import cryptobench.core.Cipher
import cryptobench.util.BlockCodec
import cryptobench.util.Bytes
import cryptobench.util.Hex

final class XteaCipher(key: XteaKey) extends Cipher {

  override def name: String = "xtea"

  override def encrypt(plaintext: String): String = {
    val data = Bytes.pad(Bytes.of(plaintext), 8)
    var off = 0
    while (off < data.length) {
      BlockCodec.write(data, off, XteaRounds.encryptBlock(BlockCodec.read(data, off), key))
      off += 8
    }
    Hex.encode(data)
  }

  override def decrypt(ciphertext: String): String = {
    val data = Hex.decode(ciphertext)
    var off = 0
    while (off < data.length) {
      BlockCodec.write(data, off, XteaRounds.decryptBlock(BlockCodec.read(data, off), key))
      off += 8
    }
    Bytes.toText(data).trim
  }
}
